js/genkit/src/genkit.ts

```diff
@@ -566,6 +566,20 @@ export class Genkit implements HasRegistry {
       } else if (opt.use) {
         resultOptions.use = opt.use;
       }
+      if ((promptResult as GenerateOptions).tools) {
+        resultOptions.tools = (promptResult as GenerateOptions).tools;
+      } else if (p.__config?.tools) {
+        resultOptions.tools = p.__config?.tools;
+      } else if (opt.tools) {
+        resultOptions.tools = opt.tools;
+      }
+      if ((promptResult as GenerateOptions).toolChoice) {
+        resultOptions.toolChoice = (promptResult as GenerateOptions).toolChoice;
+      } else if (p.__config?.toolChoice) {
+        resultOptions.toolChoice = p.__config?.toolChoice;
+      } else if (opt.toolChoice) {
+        resultOptions.toolChoice = opt.toolChoice;
+      }
```

Comment:
Wouldn't this work?

```suggestion
      resultOptions.tools = (promptResult as GenerateOptions).tools || (p.__config?.tools) || opt.tools;
      resultOptions.toolChoice = (promptResult as GenerateOptions).toolChoice || p.__config?.toolChoice || opt.toolChoice;
```

(making sure you delete undefined keys using that utility somewhere below here)

Reply:
yes, was trying to avoid undefined values... I'm just removing this whole thing in the next prompt redesign PR.